Snap 585

kneeraj opened the pull request:
During restart now every node acquires a dd readlock so no new ddl can be fired till the node is not fully up.
Also reinstalled the thread dump handler for signal URG

What is left is that will just keep the previous postCreateDDLReplay as is so that the code is there to be reactivated if found suitable later.

Commit:
With this change a new node coming up acquires a data dictionary read lock
before replaying the ddl.
The GfxdDDLMessage process returns a reply excepton to sender if it receives
a ddl to execute while the ddl replay is still going on.

Commit:
enabling signal handling for dumping stacks

Commit:
Merge branch 'snappy/master' into snap-585

Commit:
Fixing multiple junit and dunit failures (SNAP-27 SNAP-28)

 * fixing failures in FabricServerTest by setting gemfirexd.properties as GemFire
   properties file; allow for skipping "gemfirexd." properties in the GemFire layer
 * changed GemFireStore.boot to read many properties either from the incoming props
   or from gemfirexd.properties/SystemProperty with "gemfirexd." prefix; these
   include properties like host-data, server-groups, stand-alone-locator etc
 * separated the dunit test parallelization logic separately into a groovy function
   in top-level build.gradle and use the same for junit tests to reduce the running time;
   updated build.gradle targets and README.md accordingly
 * fixing more occasional failures in ExpirationDUnit, GemFireXDHADUnit
 * fixing the gemfirexd-core jar name in AnalyzeSerializablesJUnitTest
 * correcting ConnectionURLTest, XATransactionTest to use mcast-port=0 since tests run with
   gemfire.disallowMcastDefaults
 * fixing occasional unique constraint violation failures in ViewTest due to
   current time remaining same for all inserts
 * remove the gemfirexd.system.home setting in TestConfiguration that gets
   inherited into junit tests causing failure in FabricServerTest

With above changes there are failures in exactly two junits: TransactionHDFSTableTest, OffHeapTest
and failure in exactly one dunit: TransactionRRHDFSTableDUnit

There may be other occasional failures that need to be reported and fixed as seen.
Above failures will be taken care of subsequently via separate JIRA tickets.

Commit:
Merge branch 'snappy/master' into snap-585

Commit:
Fixed some issues found in precheckin

precheckin is clean now except known issues. Re-running with the latest
as now the precheckin is clean

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/store/GemFireStore.java:
```diff
@@ -2308,6 +2299,15 @@ public void setDBName(String dbname) {
+      }
+    }
   }
 
```
I think this is too late. Many a times we need this while GemFireStore is starting up (due to hangs there), so the handler should be installed upfront. I will suggest setting the "isSnappyStore" flag too upfront via an internal property passed to GemFireStore.boot, or even a system property, or set a static field in GemFireStore instead of instance field so it can be set right at the start.

kneeraj replied:
Will see.

sumwale replied:
Ideally we would like the handler to work with Spark too. Can you file a JIRA to explore if some other signal works with Spark too: SIGCHLD, SIGWINCH, SIGPWR, SIGIO. Last two may not be available on most UNIX systems, so we also need to select one which is available on the UNIX platforms we will be interested in going forward: Linux, Mac, Solaris

Commit:
Merge branch 'snappy/master' into snap-585

Commit:
Added a counter in DiskRegion recovery map encapsulating classes to keep track
of recovered TOMBSTONES and INVALID values when reading the Oplogs.
Using this while checking the sanity of indexes against the base region while recovery
as the TOMBSTONE etc. will be there in the region map but not in index's skiplist.

Commit:
Merge branch 'snappy/master' into snap-585

Commit:
Fixed a couple of issues with index re-creation.
Added a test flag to enforce index recreation for unit test.
Added a dunit to test index recreation.

Commit:
Fixed a bug where index sanity check code on restart was not ignoring stale
data

Commit:
Merge branch 'snappy/master' into snap-585

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskStoreImpl.java:
```diff
@@ -4811,16 +4814,16 @@ public void run() {
+//          for (Oplog oplog : this.allOplogs) {
+//            oplog.clearInitRecoveryMap();
+//          }
           markIndexRecoveryDone();
```
Why commented out?

kneeraj replied:
Doing it later for rebuilding. But yes after recent changes I think I cannot remove it. Will see. Anyways another pull request created for 9index sanity checks related changes. snap-588

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskStoreImpl.java:
```diff
@@ -5268,4 +5277,11 @@ else if (memoryState.isEviction()) {
+      oplog.clearInitRecoveryMap();
+    }
+  }
 }
```
Shouldn't these index/oplog related changes (recreate indexes I suppose) be part of a separate PR? This is a different issue that should have a separate JIRA associated with it, no?

kneeraj replied:
Done. snap-588 pull request.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/OplogIndex.java:
```diff
@@ -211,6 +211,12 @@ boolean checkValidIndexFile(File f) {
+      // also delete the irf of OplogIndex
+      this.irf = null;
+    }
     this.oplog.indexesWritten.clear();
```
See comments before. I think these changes should be part of a separate ticket/PR.

kneeraj replied:
Done. snap-588.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/GfxdDDLMessage.java:
```diff
@@ -151,99 +151,12 @@ protected void processMessage(final DistributionManager dm)
+    if (memStore.initialDDLReplayInProgress()) {
+      throw new ReplyException("Received ddl " + ddl + " from sender " + getSender()
+          + " while ddl replay is still in progress");
     }
```
Should throw GfxdDDLReplayInProgressException so that sender can retry/ignore otherwise the DDL will be rolled back which we do not want. I think there is a window where DD lock has not been acquired but message can still come through (which will start as soon as DD region initialization is started) so for those cases the DDL should not fail rather ignore/retry.

kneeraj replied:
Actually with rthe changes done this should never happen. I mean new ddls cannot come to a starting node as dd read lock is taken + ddl queue region initialization on this restarting node is happening after we acquire the read lock on dd so this is more of an IllegalStateException. For now replaced it with GfxdDDLReplayInProgress but inclined to throw IllegalStateException. Please suggest.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/GfxdDDLMessage.java:
```diff
@@ -151,99 +151,12 @@ protected void processMessage(final DistributionManager dm)
-        throw new GfxdDDLReplayInProgressException(
-            "skipping DDL execution due to replay in progress");
-      }
+    // With the new logic of not allowing new ddls to be fired when replay is
```
I will suggest commenting out the old code instead of removing. We don't know if customers may look for previous behaviour.

kneeraj replied:
Done as per suggestion.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -651,21 +637,7 @@ private void postCreateDDLReplay(final EmbedConnection embedConn,
-              maxIterations = 1;
-            }
-            continue;
-          }
```
I will suggest commenting out the old code instead of removing. We don't know if customers may look for previous behaviour.

kneeraj replied:
Done as per suggestion.

sumwale commented:
Neeraj, see some comments above. Assuming that you have moved the queue initialization inside postCreateDDLReplay which is not seen in this PR yet. Please do that so can take a look at that portion.

Have not looked at index related changes. I think those should be in a separate JIRA+PR.

Commit:
Moved the ddl region queue initialization to just before postCreateDDLReplay.
Changed the datadictionary read lock acquiring from current txn to null so that releaseAllLocks
later in the ddl replay does not release this read lock on data dictionary.
Added a dunit test which checks for new ddl not being fired when a node is coming up.

Commit:
Removing the index sanity related changes from this branch.

Commit:
Taking care of review comments.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -459,7 +464,32 @@ synchronized public void postCreate(
+            null, Long.MAX_VALUE / 2))) {
+          if(cnt >= 10) {
+            throw new IllegalStateException("KN: failed to acquire dd read lock");
+          }
```
Proper exception with better message. Perhaps a StandardException with SQLState.BOOT_DATABASE_FAILED.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -459,7 +464,32 @@ synchronized public void postCreate(
+        }
+        logger.info("KN: acquired dd read lock during post create");
+        // populate and initialize the DDL queue
+        if (this.memStore.restrictedDDLStmtQueue()) {
```
Log will be fine -- just remove the KN:

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -616,25 +650,26 @@ private void postCreateDDLReplay(final EmbedConnection embedConn,
+//              this.memStore, Long.MAX_VALUE / 2, true);
+//        }
+//        this.memStore.acquireDDLReplayLock(true);
+//        acquiredReplayLock = true;
```
Sorry to nag about this, but I will prefer a block comment here instead of //. Helps keep the older blame history.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -651,21 +686,22 @@ private void postCreateDDLReplay(final EmbedConnection embedConn,
+//              maxIterations = 1;
+//            }
+//            continue;
+//          }
```
Same as above about using block comments.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -681,12 +717,14 @@ private void postCreateDDLReplay(final EmbedConnection embedConn,
+//          logger.info("FabricDatabase: initial replay remaining iters "
+//              + maxIterations + " with remaining queue size " + actualSize);
+//        }
         // First check if region intialization should be skipped for
```
ditto as above

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/GfxdDDLMessage.java:
```diff
@@ -151,100 +151,110 @@ protected void processMessage(final DistributionManager dm)
+//        throw new GfxdDDLReplayInProgressException(
+//            "skipping DDL execution due to replay in progress");
+//      }
+//    }
```
ditto as above

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -987,13 +1028,14 @@ else if (qVal instanceof AbstractGfxdReplayableMessage) {
+//      }
+//      if (acquiredReplayLock) {
+//        this.memStore.releaseDDLReplayLock(true);
+//      }
```
ditto as above

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -971,10 +1011,11 @@ else if (qVal instanceof AbstractGfxdReplayableMessage) {
+//      if (ddReadLockAcquired) {
+//        this.dd.unlockAfterReading(null);
+//        ddReadLockAcquired = false;
+//      }
```
ditto as above

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -870,11 +909,12 @@ else if (qVal instanceof AbstractGfxdReplayableMessage) {
+//        // release the replay lock at this point since we will have the DD lock
+//        this.memStore.releaseDDLReplayLock(true);
+//        acquiredReplayLock = false;
+//      }
```
ditto as above

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -845,7 +883,8 @@ else if (qVal instanceof AbstractGfxdReplayableMessage) {
         }
-      }
+      // commenting out for snap-585
+      //}
```
ditto as above

sumwale commented:
Looks good. See some comments above.

Can you rebase squash your commits into a single commit? Or else better will be to try the steps outlined here to merge squash on snappy-store repo: https://wiki.apache.org/hadoop/GithubIntegration

```
git checkout snappy/master    # switch to local master
git pull --rebase             # fast-forward to current remote HEAD
git merge --squash SNAP-585
```

You can do this on the new checkout we had done for scaladocs. This will make the merge a single commit which is much cleaner and also avoid the useless separate merge commit itself.

The commit and add an appropriate commit message for the whole -- starting it with [SNAP-585] in first line will be good much like Spark. To also close the github PR automatically, have first line of commit something like:

```
[SNAP-585] Simplify DDL replay logic (fixes #36)

(rest of the description from PR here)
```

I was planning to add a script to do all this automatically while fetching the new commit message from PR's description but didn't get time yet. If you can test this it will be great. We can create appropriate guidelines and put in a script quickly.

Commit:
Took care of the review comments.
